[GTK][WPE][Skia] Add GPU synchronization primitives to NativeImage/ImageBuffer for Skia

nikolaszimmermann opened the pull request:
#### 774fd3dddd84bda499316f2f8ab5d28b42147e23
<pre>
[GTK][WPE][Skia] Add GPU synchronization primitives to NativeImage/ImageBuffer for Skia
<a href="https://bugs.webkit.org/show_bug.cgi?id=283286">https://bugs.webkit.org/show_bug.cgi?id=283286</a>

Reviewed by Said Abou-Hallawa.

Prepare for threaded GPU rendering: add Skia specific functions
to ImageBuffer(Backend)/NativeImageBackend:

 - finishAcceleratedRenderingAndCreateFence()

   Used to notify the underlying SkSurface, that it should flush
   all pending rendering operations and submit the work to the GPU.
   Afterwards inject a fence into the GL command stream, so that we
   we can wait upon completion at a later point, before attempting
   to use the ImageBuffer/NativeImage.

 - waitForAcceleratedRenderingFenceCompletion()

   Performs a &apos;serverWait()&apos; operation on the GLFence, previously
   created by finishAcceleratedRenderingAndCreateFence(). This
   instructs to GPU to await completion of the previous rendering.

finishAcceleratedRenderingAndCreateFence() will be used during
DisplayList recording, whenever an ImageBuffer/NativeImage resource
is used, upon entering the DisplayList::ResourceHeap cache.

waitForAcceleratedRenderingFenceCompletion() will be used during
DisplayList replaying, whenever an ImageBuffer/NativeImage resource
is extracted from the DisplayList::ResourceHeap cache.

This PR only contains the Skia specific ImageBuffer/NativeImage related
changes -- the hooks in the DisplayList code that make use of these
methods will be added in a separated PR.

No change in functionality for existing tests.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBuffer::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBufferBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/NativeImage.cpp:
* Source/WebCore/platform/graphics/NativeImage.h:
(WebCore::NativeImageBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::NativeImageBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::PlatformImageNativeImageBackend::waitForAcceleratedRenderingFenceCompletion):

Canonical link: <a href="https://commits.webkit.org/286767@main">https://commits.webkit.org/286767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba48b443522371408b036d56b18c7fd0bc8d43c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29832 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81470 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28202 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79043 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65103 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4254 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60283 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18360 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79993 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50229 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66038 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40597 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47631 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23533 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26528 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23864 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82906 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68556 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4456 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66011 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67810 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16931 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11799 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9880 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4249 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7062 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4272 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7703 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6030 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa reviewed (approved):
I have added some review comments related to this PR in the full patch https://github.com/WebKit/WebKit/pull/36796.

nikolaszimmermann committed:
[GTK][WPE][Skia] Add GPU synchronization primitives to NativeImage/ImageBuffer for Skia
https://bugs.webkit.org/show_bug.cgi?id=283286

Reviewed by Said Abou-Hallawa.

Prepare for threaded GPU rendering: add Skia specific functions
to ImageBuffer(Backend)/NativeImageBackend:

 - finishAcceleratedRenderingAndCreateFence()

   Used to notify the underlying SkSurface, that it should flush
   all pending rendering operations and submit the work to the GPU.
   Afterwards inject a fence into the GL command stream, so that we
   we can wait upon completion at a later point, before attempting
   to use the ImageBuffer/NativeImage.

 - waitForAcceleratedRenderingFenceCompletion()

   Performs a 'serverWait()' operation on the GLFence, previously
   created by finishAcceleratedRenderingAndCreateFence(). This
   instructs to GPU to await completion of the previous rendering.

finishAcceleratedRenderingAndCreateFence() will be used during
DisplayList recording, whenever an ImageBuffer/NativeImage resource
is used, upon entering the DisplayList::ResourceHeap cache.

waitForAcceleratedRenderingFenceCompletion() will be used during
DisplayList replaying, whenever an ImageBuffer/NativeImage resource
is extracted from the DisplayList::ResourceHeap cache.

This PR only contains the Skia specific ImageBuffer/NativeImage related
changes -- the hooks in the DisplayList code that make use of these
methods will be added in a separated PR.

No change in functionality for existing tests.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBuffer::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBufferBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/NativeImage.cpp:
* Source/WebCore/platform/graphics/NativeImage.h:
(WebCore::NativeImageBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::NativeImageBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.cpp:
(WebCore::ImageBufferSkiaAcceleratedBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::ImageBufferSkiaAcceleratedBackend::waitForAcceleratedRenderingFenceCompletion):
* Source/WebCore/platform/graphics/skia/ImageBufferSkiaAcceleratedBackend.h:
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::finishAcceleratedRenderingAndCreateFence):
(WebCore::PlatformImageNativeImageBackend::waitForAcceleratedRenderingFenceCompletion):

Canonical link: https://commits.webkit.org/286767@main

alexgcastro reviewed (approved):
LGTM!
